### ISSUE TITLE
`gelu` will backward crash

### ISSUE THREAD

**Issue by upupyy** [pytorch/pytorch#78152]
### 🐛 Describe the bug

`gelu` will backward crash
```python
import torch

input = torch.rand([0], dtype=torch.float32).requires_grad_()
res = torch.nn.functional.gelu(input, )

res.sum().backward()
# Segmentation fault (core dumped)
```

I think the reason is that pytorch tries to compute the gradient with empty output, and #78149 has the same reason

### Versions

pytorch: 1.11.0
Ubuntu: 20.04
Python: 3.9.5

cc @ezyang @gchanan @zou3519 @albanD @gqchen @pearu @nikitaved @soulitzer @Lezcano @Varal7 @gujinghui @PenghuiCheng @XiaobingSuper @jianyuh @VitalyFedyunin

**Comment by yuguo68** [comment]
thank you @upupyy for reporting these issues. I tried to reproduce on master, (1), the `embedding` code will throw an error `INDICES element is out of DATA bounds, id=-1 axis_dim=4`. If I change to `input = torch.randint(0, 3, [1, 11], dtype=torch.int64)`, it runs fine. (2) `gelu` code runs fine on master.

**Comment by upupyy** [comment]
Hi @yuguo68, for `gelu`, it still crashes on the latest version nightly pytorch

```py
import torch

input = torch.rand([0], dtype=torch.float32).requires_grad_()
res = torch.nn.functional.gelu(input, )

res.sum().backward()
#  floating point exception (core dumped)
```

The machine info:
```
PyTorch version: 1.13.0.dev20220906
Is debug build: False
CUDA used to build PyTorch: 11.3
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.4 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
Clang version: Could not collect
CMake version: version 3.16.3
Libc version: glibc-2.31

Python version: 3.9.12 (main, Jun  1 2022, 11:38:51)  [GCC 7.5.0] (64-bit runtime)
Python platform: Linux-5.4.0-117-generic-x86_64-with-glibc2.31
Is CUDA available: True
CUDA runtime version: Could not collect
GPU models and configuration:
GPU 0: NVIDIA RTX A6000
GPU 1: NVIDIA RTX A6000
GPU 2: NVIDIA RTX A6000
GPU 3: NVIDIA RTX A6000

Nvidia driver version: 510.68.02
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.2.2
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.2.2
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.2.2
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.2.2
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.2.2
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.2.2
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.2.2
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.22.3
[pip3] torch==1.13.0.dev20220906
[pip3] torchaudio==0.13.0.dev20220906
[pip3] torchvision==0.14.0.dev20220906
[conda] blas                      1.0                         mkl
[conda] cudatoolkit               11.3.1               h2bc3f7f_2
[conda] mkl                       2021.4.0           h06a4308_640
[conda] mkl-service               2.4.0            py39h7f8727e_0
[conda] mkl_fft                   1.3.1            py39hd3c417c_0
[conda] mkl_random                1.2.2            py39h51133e4_0
[conda] numpy                     1.22.3           py39he7a7128_0
[conda] numpy-base                1.22.3           py39hf524024_0
[conda] pytorch                   1.13.0.dev20220906 py3.9_cuda11.3_cudnn8.3.2_0    pytorch-nightly
[conda] pytorch-mutex             1.0                        cuda    pytorch-nightly
[conda] torchaudio                0.13.0.dev20220906      py39_cu113    pytorch-nightly
[conda] torchvision               0.14.0.dev20220906      py39_cu113    pytorch-nightly
```

**Comment by Lezcano** [comment]
I can repro

**Comment by zou3519** [comment]
Weirdly I am able to reproduce this on a nightly binary but not on a build from source of PyTorch

**Comment by zou3519** [comment]
Appears to be a mkldnn problem. Workaround is to disable mkldnn. @XiaobingSuper how do you think we should proceed here?
```py
# Workaround: disable mkldnn
with torch.backends.mkldnn.flags(enabled=False):
    input = torch.rand([0], dtype=torch.float32).requires_grad_()
    res = torch.nn.functional.gelu(input, )
    res.sum().backward()
```

**Comment by Lezcano** [comment]
IMO we could simply do something like:
```c++
if (input.numel() == 0) {
  return at::zeros_like(input);
}
```

We do this a lot in linalg functions, as some of MAGMA / cuBLAS / cuSOLVER etc,  don't support empty inputs.

**Comment by XiaobingSuper** [comment]
@Lezcano , I think we can disable the OneDN path for empty inputs.
